File: Problems/0399-evaluate-division.py
from collections import defaultdict, deque

class Solution(object):
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        graph = defaultdict(list)

        for (a, b), v in zip(equations, values):
            graph[a].append((b, v))
            graph[b].append((a, 1.0 / v))

        ans = []

        for src, dst in queries:
            if src not in graph or dst not in graph:
                ans.append(-1.0)
                continue

            if src == dst:
                ans.append(1.0)
                continue

            q = deque([(src, 1.0)])
            vis = {src}
            found = -1.0

            while q:
                node, val = q.popleft()

                if node == dst:
                    found = val
                    break

                for nxt, w in graph[node]:
                    if nxt not in vis:
                        vis.add(nxt)
                        q.append((nxt, val * w))

            ans.append(found)

        return ans

File: Problems/0399-evaluate-division.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations,
                                 vector<double>& values,
                                 vector<vector<string>>& queries) {
        unordered_map<string, vector<pair<string, double>>> graph;

        for (int i = 0; i < equations.size(); i++) {
            string a = equations[i][0];
            string b = equations[i][1];
            double v = values[i];

            graph[a].push_back({b, v});
            graph[b].push_back({a, 1.0 / v});
        }

        vector<double> ans;

        for (auto &q : queries) {
            string src = q[0], dst = q[1];

            if (!graph.count(src) || !graph.count(dst)) {
                ans.push_back(-1.0);
                continue;
            }

            if (src == dst) {
                ans.push_back(1.0);
                continue;
            }

            queue<pair<string, double>> bfs;
            unordered_set<string> vis;

            bfs.push({src, 1.0});
            vis.insert(src);

            double result = -1.0;

            while (!bfs.empty()) {
                auto [node, value] = bfs.front();
                bfs.pop();

                if (node == dst) {
                    result = value;
                    break;
                }

                for (auto &[next, weight] : graph[node]) {
                    if (!vis.count(next)) {
                        vis.insert(next);
                        bfs.push({next, value * weight});
                    }
                }
            }

            ans.push_back(result);
        }

        return ans;
    }
};

